topi/python/topi/cuda/conv2d_int8.py

```diff
@@ -0,0 +1,336 @@
+    cfg.define_split("tile_f", cfg.axis(f), num_outputs=4)
+    cfg.define_split("tile_y", cfg.axis(y), num_outputs=4)
+    cfg.define_split("tile_x", cfg.axis(x), num_outputs=4)
+
+    bf, vf, tf, fi = cfg["tile_f"].apply(s, output, f)
+    by, vy, ty, yi = cfg["tile_y"].apply(s, output, y)
+    bx, vx, tx, xi = cfg["tile_x"].apply(s, output, x)
+
+    # this is the scope to attach global config inside this kernel
+    kernel_scope, n = s[output].split(n, nparts=1)
+
+    max_block_z = 128
+    if batch > max_block_z:
+        _, n = s[output].split(n, factor=max_block_z)
+    s[output].reorder(n, bf, by, bx, vf, vy, vx, tf, ty, tx, fi, yi, xi)
+    fused_byx = s[output].fuse(by, bx)
+    s[output].bind(n, tvm.thread_axis("blockIdx.z"))
+    s[output].bind(bf, tvm.thread_axis("blockIdx.y"))
+    s[output].bind(fused_byx, tvm.thread_axis("blockIdx.x"))
+    s[output].bind(vf, tvm.thread_axis("vthread"))
+    s[output].bind(vy, tvm.thread_axis("vthread"))
+    s[output].bind(vx, tvm.thread_axis("vthread"))
+    s[output].bind(tf, tvm.thread_axis("threadIdx.z"))
+    s[output].bind(ty, tvm.thread_axis("threadIdx.y"))
+    s[output].bind(tx, tvm.thread_axis("threadIdx.x"))
+
+    s[conv].compute_at(s[output], tx)
+
+    # tile and bind reduction axes
+    n, f, y, x, c = s[conv].op.axis
+
+    rc, ry, rx, rc_block = s[conv].op.reduce_axis
+    cfg.define_split("tile_rc", cfg.axis(rc), num_outputs=2)
+    cfg.define_split("tile_ry", cfg.axis(ry), num_outputs=2)
+    cfg.define_split("tile_rx", cfg.axis(rx), num_outputs=2)
+    rco, rci = cfg['tile_rc'].apply(s, conv, rc)
+    ryo, ryi = cfg['tile_ry'].apply(s, conv, ry)
+    rxo, rxi = cfg['tile_rx'].apply(s, conv, rx)
+
+    s[conv].reorder(rco, ryo, rxo, rci, ryi, rxi, n, f, y, x, c, rc_block)
+
+    _, rc_block = s[conv].split(rc_block, factor=4)
+    s[conv].tensorize(rc_block, _dp4a)
+
+    s[AA].compute_at(s[conv], rxo)
+    s[WW].compute_at(s[conv], rxo)
+
+    # cooperative fetching
+    for load in [AA, WW]:
+        if load == AA:
+            n, f, y, x, c = s[load].op.axis
+            if pad_data == packed_data and stride_h == 1 and stride_w == 1:
+                s[load].vectorize(c)
+                fused = s[load].fuse(n, f, y, x)
+            else:
+                c, _ = s[load].split(c, factor=4)
+                fused = s[load].fuse(n, f, y, x, c)
+        else:
+            n, f, y, x, oc_chunk, c = s[load].op.axis
+            fused = s[load].fuse(n, f, y, x, oc_chunk)
+            s[load].vectorize(c)
+
+        fused, tx = s[load].split(fused, factor=cfg["tile_x"].size[2])
+        fused, ty = s[load].split(fused, factor=cfg["tile_y"].size[2])
+        fused, tz = s[load].split(fused, factor=cfg["tile_f"].size[2])
+        s[load].bind(tz, tvm.thread_axis("threadIdx.z"))
+        s[load].bind(ty, tvm.thread_axis("threadIdx.y"))
+        s[load].bind(tx, tvm.thread_axis("threadIdx.x"))
+
+    # double buffer
+    cfg.define_knob('AA_double_buffer', [0, 1])
```

Comment:
Just curious, are there cases where double buffering helped?

Reply:
I checked tuned result of AutoTVM. In vgg, most layers use double buffering. But in resnet, it seems this doesn't help. Some layers use double buffering, while others do not.